A48-xds-least-request-lb-policy.md

```diff
@@ -0,0 +1,225 @@
+        continue;
+    }
+
+    if (sampled.active_requests < candidate.active_requests) {
+        candidate = sampled;
+    }
+}
+return candidate;
+```
+
+This pseudo-code is mirroring the
+[Envoy implementation](https://github.com/envoyproxy/envoy/blob/2443032526cf6e50d63d35770df9473dd0460fc0/source/common/upstream/load_balancer_impl.cc#L844-L865)
+
+##### Outstanding request counters
+
+The `least_request_experimental` policy will associate one outstanding request counter for each subchannel.
+These counters should additionally share the lifecycle with its corresponding subchannel.
+The counter for a subchannel should be atomically incremented by one after it has been successfully
+picked by the picker. Due to language-specifics, the increment may occur either before or during stream creation.
+In the `PickResult` for each language-specific implementation (
+[Java](https://github.com/grpc/grpc-java/blob/1f90e0e28d5628195cb1f861b73e45ed003f2973/api/src/main/java/io/grpc/LoadBalancer.java#L561-L571),
+[C++](https://github.com/grpc/grpc/blob/4567af504ed53cc8398e53bf1efd23f753d43bb8/src/core/ext/filters/client_channel/lb_policy.h#L178-L201),
+[Go](https://github.com/grpc/grpc-go/blob/01ed64857e3146000ec99cdea4f2932204f17cdd/balancer/balancer.go#L256-L269)
+), the picker should add a
+callback for atomically decrementing the subchannel counter once the RPC finishes (regardless of `Status` code).
+This approach closely resembles the implementation for outstanding request counters already present in
+the `cluster_impl_experimental` policy (
+[Java](https://github.com/grpc/grpc-java/blob/0000cba665c69958355b639474c7387d98afcc79/xds/src/main/java/io/grpc/xds/ClusterImplLoadBalancer.java#L282-L379),
+[C++](https://github.com/grpc/grpc/blob/79d684529d9db60aa2c5e82deb43e297a6818cfb/src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_impl.cc#L285-L354),
+[Go](https://github.com/grpc/grpc-go/blob/03268c8ed29e801944a2265a82f240f7c0e1b1c3/xds/internal/balancer/clusterimpl/picker.go#L102-L191)
+).
+This approach entails some degree of raciness which will be discussed later
+(See [Outstanding request counter raciness](#outstanding-request-counter-raciness)).
+
+### Temporary environment variable protection
+
+During initial development, this feature will be enabled via the
+`GRPC_XDS_EXPERIMENTAL_ENABLE_LEAST_REQUEST` environment variable.  This
+environment variable protection will be removed once the feature has
+proven stable.
+
+## Rationale
+
+### Support for Weighted Least Request
+
+While also supporting the "*all weights not equal*" (Weighted Least Request) case
+[from Envoy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#weighted-least-request)
+would enable more advanced use cases, there are some missing pieces in gRPC that
+are required for that to be technically feasible.
+For example, gRPC ignores the per-endpoint `load_balancing_weight` field as a result
```

Comment:
We actually added support for this field in C-core as part of implementing `ring_hash`.  I don't think it would be hard to support in other languages either.

What would be harder would be actually changing `ROUND_ROBIN` to support these weights.  There is another open gRFC proposal for that in #202, although it's been silent for a while, so I'm not sure if the author is still working on it.

Reply:
Interesting!
Looks like java at least supports it as well but the weight attribute is marked as:
```
Internal attributes used for xDS implementation. Do not use.
```
which means per-endpoint weights still may need to be standardized in a gRFC before it can be used in an LB implementation like this.
Will update the text to reflect this.

Reply:
Just FYI, in C-core, we added the attribute in a way that is not xDS-specific:

https://github.com/grpc/grpc/blob/c1089d2964528b52a25c18d748910658e4762d40/src/core/ext/filters/client_channel/server_address.h#L116

Note that I think we'd view this more as a convention than as a required part of the API: it's up to each resolver impl to decide to add this attribute to the addresses it produces, and it's up to each LB policy impl to decide whether to look at this attribute and what to do with it.  So it's something that can be used when there are use-cases for which it makes sense, but it's not required or assumed to be used by any given resolver or LB policy.

But I think within those constraints, we could support it in `least_request`, the same way that we do in `ring_hash`.

Reply:
Been thinking about this a bit more and I'm currently contemplating whether to include the "all weights not equal" (weighted least request) in this proposal or not.
The main reasons I had for not including it where:
1) I didn't want to deal with design of sending per-endpoint weights to LBs in each language in a non-xds-specific way.
2) I have a hard time seeing how the control plane can calculate valuable per-endpoints weights in cases where endpoints are considered equal (at least in terms of cpu/mem performance). The control plane would probably need to have per-endpoint load information either through load reports or other means. And even in that case the feedback loop may be too slow for large scale dynamic setups with high/bursty rps. This is where ORCA may be beneficial instead.

If you are saying I wouldn't have to think about 1) in this proposal, it all boils down to whether the feature is worth the moderately extra effort to implement or not.
If you think it's worth the effort, I can add the extra picker logic for the "all weights not equal" case and state that it's "optional for a language to implement given a non-xds-specific solution for passing the weights from the resolver".

Reply:
I don't have any strong feeling one way or the other about whether the "all requests not equal" algorithm is useful.  The only factor that I would consider here is forward-compatibility: if it's not supported now and then we wind up adding it later, will clients start behaving differently when they get upgraded?  If that change is something that might cause problems, then it might make sense to implement it now, just so that we don't run into that later.  OTOH, if that's purely an optimization and not likely to cause any problems, then it may be fine to add it later if/when someone actually needs it.

Reply:
I think we would prefer to not include the "all weights not equal" algorithm in this proposal. Even though I think the implementation doesn't need to be too tricky, there are still some tricky edge cases which probably need to be discussed/addressed.
As for the consequences of adding this to the implementation later, it seems like the behavioral change wouldn't be too severe/controversial, especially since it only would affect the case where the control plane is sending per-endpoint weights to a proxyless gRPC client which previously ignored them.
This exact problem already exists for round_robin as well, since the ROUND_ROBIN policy in xDS should mean weighted round robin if per-endpoint weights are present. But until https://github.com/grpc/proposal/pull/202 (or similar) is implemented, proxyless gRPC will ignore any per-endpoint weights for a cluster with ROUND_ROBIN policy.

I might also be willing to open a follow-up gRFC for this addition to least request, but it's currently not a priority for us to have it supported.

Reply:
Okay, sounds good.  I'm fine with not including the "all weights not equal" algorithm for now.